lib/rubyboy/bus.rb

```diff
@@ -11,74 +11,93 @@ def initialize(ppu, rom, ram, mbc, timer, interrupt, joypad, apu)
       @apu = apu
       @interrupt = interrupt
       @timer = timer
+    end
+
+    def read_byte(addr)
+      case addr >> 12
+      when 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0xa, 0xb
+        return @mbc.read_byte(addr)
+      when 0x8, 0x9
+        return @ppu.read_byte(addr)
+      when 0xc
+        return @ram.wram1[addr - 0xc000]
+      when 0xd
+        return @ram.wram2[addr - 0xd000]
+      when 0xf
+        case (addr >> 8) & (0xf)
+        when 0xe
+          return @ppu.read_byte(addr) if addr <= 0xfe9f
+        when 0xf
+          last_byte = addr & 0xFF
 
-      @read_methods = Array.new(0x10000)
-      @write_methods = Array.new(0x10000)
+          case last_byte
+          when 0x00
+            return @joypad.read_byte(addr)
+          when 0x04, 0x05, 0x06, 0x07
+            return @timer.read_byte(addr)
+          when 0x0f
+            return @interrupt.read_byte(addr)
+          when 0x46
+            return @ppu.read_byte(addr)
+          when 0xff
+            return @interrupt.read_byte(addr)
+          end
 
-      set_methods
-    end
+          return @apu.read_byte(addr) if last_byte <= 0x26 && last_byte >= 0x10
```

Comment:
Just a minor detail, but since it was originally written as `when 0xff10..0xff26`, it might be more readable to write it starting with the smaller value, like `if last_byte >= 0x10 && last_byte <= 0x26`.

Reply:
It was done this way for performance reasons. Almost all values are `>= 0x10` few `<= 0x26`, so it avoid a comparison in most cases.

If Ruby had a very advanced optimizing compiler that wouldn't be needed, but it's not the case yet.

But if you fell very strongly about it, I can change that.

Reply:
I see, got it!

I don't have a strong opinion, so it's fine as is.